File: services/routers/invio_commercialista.py
"""Invio degli XML al commercialista — l'area admin (fase D del piano, 25/09/2026).

Dalla scheda cliente l'admin:
  - collega una P.IVA del cliente a Invoicetronic: il company_id si CERCA
    (`GET /company/IT<piva>`), si confronta con quello delle fatture gia'
    arrivate e si salva solo se coincide con quello che l'admin ha visto;
  - scrive email del commercialista, frequenza, data di partenza, e registra il
    consenso (vale per l'email di quel momento);
  - lancia la prova a vuoto, l'invio (il primo o, dopo, quello del periodo
    maturato) e i reinvii;
  - chiarisce gli esiti incerti e annulla una richiesta non ancora presa.

Nessun endpoint spedisce: scrivono una riga `richiesto` nel registro, che
l'esecutore del queue-worker prende entro un minuto. E' il DB (trigger e vincoli
della migration 20260925113943) a rifiutare cio' che non e' autorizzato: qui i
suoi rifiuti diventano frasi per l'admin.
"""
from __future__ import annotations

import logging
import re
from datetime import date, timedelta
from typing import Any, Dict, List, Literal, Optional, Set

from fastapi import APIRouter, Depends, Header, HTTPException
from pydantic import BaseModel

from services import invio_commercialista_service as svc
from services.routers.admin import _verify_admin
from utils.supabase_paging import fetch_all


def _fw():
    import services.fastapi_worker as fw
    return fw


def get_supabase_client(*args, **kwargs):
    return _fw().get_supabase_client(*args, **kwargs)


def _verify_worker_key(x_worker_key: Optional[str] = Header(None)) -> None:
    return _fw()._verify_worker_key(x_worker_key)


def _client_invoicetronic():
    from services.invoicetronic_client import ClientInvoicetronic
    return ClientInvoicetronic()


router = APIRouter(dependencies=[Depends(_verify_worker_key)])
logger = logging.getLogger("fastapi_worker")

BASE = "/api/admin/clienti/{cliente_id}/invio-commercialista"
_PIVA = re.compile(r"^[0-9]{11}$")
DUE_ANNI = (
    "Invoicetronic conserva le fatture ricevute per 2 anni. Per periodi precedenti "
    "usa il Cassetto fiscale dell'Agenzia delle Entrate."
)
# I vincoli della migration, per nome: (status, frase per l'admin).
_VINCOLI = {
    "icc_attivabile_chk": (400, "Per attivare servono l'email del commercialista, il consenso per quell'email, "
                                "la data di partenza e il collegamento a Invoicetronic."),
    "icc_email_chk": (400, "Email non valida."),
    "icc_frequenza_chk": (400, "Frequenza non valida."),
    "ici_mai_oggi_chk": (400, "Il periodo deve finire al piu' tardi ieri."),
    "ici_due_anni_chk": (400, DUE_ANNI),
    "ici_periodo_chk": (400, "La data di inizio viene dopo quella di fine."),
    "icc_piva_unica": (409, "Per questa P.IVA esiste gia' una configurazione."),
    "icc_company_unica": (409, "Questa azienda Invoicetronic e' gia' collegata a un'altra configurazione."),
    "ici_uno_in_volo": (409, "C'e' gia' un invio in corso o da chiarire per questo cliente: aspetta che finisca."),
    "ici_un_solo_primo": (409, "Il primo invio e' gia' stato fatto."),
    "ici_inizio_unico": (409, "Quel periodo e' gia' stato inviato o e' in corso."),
}


def _oggi() -> date:
    return svc.a_roma(svc.adesso_utc()).date()


def _rifiuto(exc: Exception) -> HTTPException:
    """Il DB ha detto no: la frase giusta per l'admin. I messaggi dei trigger sono
    gia' frasi in italiano senza dati personali; i vincoli si traducono per nome."""
    messaggio = str(getattr(exc, "message", "") or exc).strip()
    for nome, (status, frase) in _VINCOLI.items():
        if f'"{nome}"' in messaggio:
            return HTTPException(status_code=status, detail=frase)
    codice = getattr(exc, "code", None)
    if codice == "23P01":
        return HTTPException(status_code=409, detail="Il periodo si sovrappone a un invio gia' fatto o in corso.")
    if codice in ("23514", "23505"):
        return HTTPException(status_code=400, detail=messaggio.splitlines()[0] if messaggio else "Richiesta rifiutata")
    raise exc


def _cliente(sb, cliente_id: str) -> None:
    if not sb.table("users").select("id").eq("id", cliente_id).limit(1).execute().data:
        raise HTTPException(status_code=404, detail="Cliente non trovato")


def _configurazione(sb, cliente_id: str, config_id: str) -> Dict[str, Any]:
    righe = (
        sb.table(svc.CONFIG).select("*")
        .eq("id", config_id).eq("user_id", cliente_id).limit(1).execute().data
    )
    if not righe:
        raise HTTPException(status_code=404, detail="Configurazione non trovata")
    return righe[0]


def _piva_del_cliente(sb, cliente_id: str) -> List[str]:
    sedi = fetch_all(sb.table("ristoranti").select("id,partita_iva").eq("user_id", cliente_id).order("id"))
    registrate = fetch_all(sb.table("piva_ristoranti").select("id,piva").eq("user_id", cliente_id).order("id"))
    tutte = {r.get("partita_iva") for r in sedi} | {r.get("piva") for r in registrate}
    return sorted(p for p in tutte if isinstance(p, str) and _PIVA.match(p))


def _company_gia_viste(sb, piva: str) -> Set[int]:
    """I company_id che il webhook ha gia' scritto per questa P.IVA."""
    righe = fetch_all(
        sb.table("fatture_queue").select("id,payload_meta")
        .eq("source", "invoicetronic").eq("piva_raw", piva).order("id")
    )
    viste = set()
    for riga in righe:
        valore = (riga.get("payload_meta") or {}).get("invoicetronic_company_id")
        if isinstance(valore, int) and not isinstance(valore, bool):
            viste.add(valore)
    return viste


def _trova_azienda(sb, cliente_id: str, piva: str) -> Dict[str, Any]:
    from services.invoicetronic_client import ErroreInvoicetronic

    if not _PIVA.match(piva or "") or piva not in _piva_del_cliente(sb, cliente_id):
        raise HTTPException(status_code=400, detail="La P.IVA non e' di una sede di questo cliente")
    try:
        azienda = _client_invoicetronic().azienda_per_piva(piva)
    except ErroreInvoicetronic as exc:
        raise HTTPException(status_code=503, detail=f"Invoicetronic non risponde come dovrebbe: {exc}") from None
    if azienda is None:
        raise HTTPException(
            status_code=404,
            detail="Su Invoicetronic questa P.IVA non c'e' ancora: l'azienda nasce all'arrivo della prima "
                   "fattura sul codice destinatario di OneFlux.",
        )
    viste = _company_gia_viste(sb, piva)
    if viste and viste != {azienda["id"]}:
        raise HTTPException(
            status_code=409,
            detail=f"Invoicetronic risponde con l'azienda {azienda['id']}, ma le fatture gia' arrivate per "
                   f"questa P.IVA vengono da {sorted(viste)}: non si collega, va chiarito col supporto.",
        )
    return {"company_id": azienda["id"], "nome": azienda.get("name"), "vat": azienda.get("vat"),
            "gia_viste": sorted(viste)}


# ── Letture ─────────────────────────────────────────────────────────────────

_COLONNE_INVII = (
    "id,tipo,stato,periodo_dal,periodo_al,richiesto_da,n_file,byte_totali,destinatario,motivo,"
    "creata_at,iniziata_at,conclusa_at,email_tentata_at,brevo_http_status,link_scade_il,"
    "file_rimossi_at,chiarito_non_partito_at"
)


@router.get(BASE, tags=["Admin"])
def invio_commercialista_stato(cliente_id: str, admin_user: dict = Depends(_verify_admin)) -> Dict[str, Any]:
    sb = get_supabase_client()
    _cliente(sb, cliente_id)
    oggi = _oggi()
    configurazioni = fetch_all(sb.table(svc.CONFIG).select("*").eq("user_id", cliente_id).order("creata_at"))
    risultato = []
    for config in configurazioni:
        invii = (
            sb.table(svc.INVII).select(_COLONNE_INVII)
            .eq("config_id", config["id"]).order("creata_at", desc=True).limit(20).execute().data or []
        )
        ultimo = svc.ultimo_giorno_inviato(sb, config["id"])
        risultato.append({
            **config,
            "sede_sdi_attiva": svc.sede_sdi_attiva(sb, cliente_id, config["piva"]),
            "ultimo_giorno_inviato": ultimo.isoformat() if ultimo else None,
            "invii": invii,
        })
    collegate = {c["piva"] for c in configurazioni}
    return {
        "oggi": oggi.isoformat(),
        "limite_due_anni": svc.limite_due_anni(oggi).isoformat(),
        "piva_disponibili": [p for p in _piva_del_cliente(sb, cliente_id) if p not in collegate],
        "configurazioni": risultato,
    }


@router.get(BASE + "/azienda", tags=["Admin"])
def invio_commercialista_cerca_azienda(
    cliente_id: str, piva: str, admin_user: dict = Depends(_verify_admin),
) -> Dict[str, Any]:
    sb = get_supabase_client()
    _cliente(sb, cliente_id)
    return _trova_azienda(sb, cliente_id, piva.strip())


# ── Configurazione ──────────────────────────────────────────────────────────

class CollegaBody(BaseModel):
    piva: str
    company_id: int


@router.post(BASE, tags=["Admin"])
def invio_commercialista_collega(
    cliente_id: str, body: CollegaBody, admin_user: dict = Depends(_verify_admin),
) -> Dict[str, Any]:
    """Crea la configurazione (spenta). Il company_id lo decide Invoicetronic,
    non il body: il body dice solo quale azienda l'admin ha visto e confermato."""
    sb = get_supabase_client()
    _cliente(sb, cliente_id)
    azienda = _trova_azienda(sb, cliente_id, body.piva.strip())
    if azienda["company_id"] != body.company_id:
        raise HTTPException(status_code=409, detail="Invoicetronic ora risponde con un'altra azienda: ricarica e ricontrolla.")
    try:
        righe = sb.table(svc.CONFIG).insert({
            "user_id": cliente_id,
            "piva": body.piva.strip(),
            "invoicetronic_company_id": azienda["company_id"],
            "invoicetronic_nome": azienda["nome"],
        }).execute().data
    except HTTPException:
        raise
    except Exception as exc:
        raise _rifiuto(exc) from None
    logger.info("invio_commercialista: collegata P.IVA per cliente %s | admin=%s",
                    cliente_id, admin_user.get("email"))
    return righe[0]


class ModificaBody(BaseModel):
    email_destinatario: Optional[str] = None
    frequenza: Optional[Literal["settimanale", "quindicinale", "mensile"]] = None
    data_partenza: Optional[date] = None
    consenso_data: Optional[date] = None
    revoca_consenso: bool = False
    attivo: Optional[bool] = None
    riprendi: bool = False


@router.patch(BASE + "/{config_id}", tags=["Admin"])
def invio_commercialista_modifica(
    cliente_id: str, config_id: str, body: ModificaBody, admin_user: dict = Depends(_verify_admin),
) -> Dict[str, Any]:
    """In tre passi, perche' il trigger azzera consenso e attivazione quando
    cambia l'email: prima i dati, poi il consenso (per l'email appena scritta),
    poi accensione o ripresa."""
    sb = get_supabase_client()
    config = _configurazione(sb, cliente_id, config_id)
    oggi = _oggi()

    dati: Dict[str, Any] = {}
    if body.email_destinatario is not None:
        email = body.email_destinatario.strip().lower()
        dati["email_destinatario"] = email or None
    if body.frequenza is not None:
        dati["frequenza"] = body.frequenza
    if body.data_partenza is not None:
        if body.data_partenza < svc.limite_due_anni(oggi):
            raise HTTPException(status_code=400, detail=DUE_ANNI)
        if body.data_partenza >= oggi:
            raise HTTPException(status_code=400, detail="La data di partenza deve essere passata.")
        dati["data_partenza"] = body.data_partenza.isoformat()
    if body.revoca_consenso:
        dati.update({"consenso_ricevuto": False, "consenso_data": None, "consenso_email": None, "attivo": False})

    passi: List[Dict[str, Any]] = [dati] if dati else []
    if body.consenso_data is not None:
        if body.consenso_data > oggi:
            raise HTTPException(status_code=400, detail="Il consenso non puo' avere una data futura.")
        email = dati.get("email_destinatario", config.get("email_destinatario"))
        if not email:
            raise HTTPException(status_code=400, detail="Prima scrivi l'email del commercialista: il consenso vale per quell'email.")
        passi.append({"consenso_ricevuto": True, "consenso_data": body.consenso_data.isoformat(), "consenso_email": email})
    finale: Dict[str, Any] = {}
    if body.attivo is not None:
        finale["attivo"] = body.attivo
    if body.riprendi:
        finale.update({"sospesa_at": None, "sospesa_motivo": None})
    if finale:
        passi.append(finale)
    if not passi:
        raise HTTPException(status_code=400, detail="Nessun campo da aggiornare")

    for passo in passi:
        try:
            sb.table(svc.CONFIG).update(passo).eq("id", config_id).eq("user_id", cliente_id).execute()
        except Exception as exc:
            raise _rifiuto(exc) from None
    logger.info("invio_commercialista: configurazione %s aggiornata (%s) | admin=%s",
                    config_id[:8], sorted(k for p in passi for k in p), admin_user.get("email"))
    return _configurazione(sb, cliente_id, config_id)


# ── Invii ───────────────────────────────────────────────────────────────────

class InvioBody(BaseModel):
    tipo: Literal["prova", "invia_ora", "reinvio"]
    dal: Optional[date] = None
    al: Optional[date] = None


@router.post(BASE + "/{config_id}/invii", tags=["Admin"])
def invio_commercialista_richiedi(
    cliente_id: str, config_id: str, body: InvioBody, admin_user: dict = Depends(_verify_admin),
) -> Dict[str, Any]:
    """Scrive la richiesta. «Invia ora» e' il primo invio (dalla data di partenza)
    finche' non ce n'e' uno riuscito, poi il periodo maturato dall'ultimo."""
    sb = get_supabase_client()
    config = _configurazione(sb, cliente_id, config_id)
    if config.get("invoicetronic_company_id") is None:
        raise HTTPException(status_code=400, detail="Prima collega la P.IVA a Invoicetronic.")
    oggi = _oggi()
    ieri, limite = oggi - timedelta(days=1), svc.limite_due_anni(oggi)

    if body.tipo == "invia_ora":
        ultimo = svc.ultimo_giorno_inviato(sb, config_id)
        if ultimo is None:
            if not config.get("data_partenza"):
                raise HTTPException(status_code=400, detail="Manca la data di partenza.")
            tipo, dal = "primo", max(date.fromisoformat(str(config["data_partenza"])[:10]), limite)
        else:
            tipo, dal = "ordinario", max(ultimo + timedelta(days=1), limite)
        al = ieri
        if dal > al:
            raise HTTPException(status_code=400, detail="Niente da inviare: fino a ieri e' gia' stato tutto inviato.")
    else:
        # Fine entro ieri, inizio entro 2 anni, inizio prima della fine: li
        # controllano i vincoli del registro, e _rifiuto li dice con queste parole.
        if body.dal is None or body.al is None:
            raise HTTPException(status_code=400, detail="Indica il periodo (dal, al).")
        tipo, dal, al = body.tipo, body.dal, body.al

    if tipo != "prova" and not svc.sede_sdi_attiva(sb, cliente_id, config["piva"]):
        raise HTTPException(status_code=400, detail="Sospesa: nessuna sede di questa P.IVA ha la ricezione SDI attiva.")
    try:
        righe = sb.table(svc.INVII).insert({
            "config_id": config_id,
            "user_id": cliente_id,
            "piva": config["piva"],
            "invoicetronic_company_id": config["invoicetronic_company_id"],
            "destinatario": config.get("email_destinatario"),
            "tipo": tipo,
            "periodo_dal": dal.isoformat(),
            "periodo_al": al.isoformat(),
            "stato": "richiesto",
            "richiesto_da": "admin",
        }).execute().data
    except Exception as exc:
        raise _rifiuto(exc) from None
    logger.info("invio_commercialista: richiesto %s %s..%s per configurazione %s | admin=%s",
                    tipo, dal, al, config_id[:8], admin_user.get("email"))
    return righe[0]


def _invio(sb, cliente_id: str, config_id: str, invio_id: str) -> None:
    _configurazione(sb, cliente_id, config_id)
    righe = (
        sb.table(svc.INVII).select("id")
        .eq("id", invio_id).eq("config_id", config_id).limit(1).execute().data
    )
    if not righe:
        raise HTTPException(status_code=404, detail="Invio non trovato")


class ChiarisciBody(BaseModel):
    esito: Literal["arrivata", "non_arrivata"]


@router.post(BASE + "/{config_id}/invii/{invio_id}/chiarisci", tags=["Admin"])
def invio_commercialista_chiarisci(
    cliente_id: str, config_id: str, invio_id: str, body: ChiarisciBody,
    admin_user: dict = Depends(_verify_admin),
) -> Dict[str, Any]:
    """Un esito incerto si chiude guardando i log di Brevo. «Non arrivata» libera
    il periodo: il prossimo invio lo rispedisce."""
    sb = get_supabase_client()
    _invio(sb, cliente_id, config_id, invio_id)
    adesso = svc.adesso_utc().isoformat()
    if body.esito == "arrivata":
        campi = {"stato": "inviato", "motivo": "chiarito dall'admin: arrivata", "conclusa_at": adesso}
    else:
        campi = {"stato": "errore", "chiarito_non_partito_at": adesso,
                 "motivo": "chiarito dall'admin: non arrivata", "conclusa_at": adesso}
    try:
        righe = (
            sb.table(svc.INVII).update(campi)
            .eq("id", invio_id).eq("stato", "esito_incerto").execute().data
        )
    except Exception as exc:
        raise _rifiuto(exc) from None
    if not righe:
        raise HTTPException(status_code=409, detail="Si chiarisce solo un invio dall'esito incerto.")
    logger.info("invio_commercialista: invio %s chiarito (%s) | admin=%s",
                    invio_id[:8], body.esito, admin_user.get("email"))
    return righe[0]


@router.post(BASE + "/{config_id}/invii/{invio_id}/annulla", tags=["Admin"])
def invio_commercialista_annulla(
    cliente_id: str, config_id: str, invio_id: str, admin_user: dict = Depends(_verify_admin),
) -> Dict[str, Any]:
    sb = get_supabase_client()
    _invio(sb, cliente_id, config_id, invio_id)
    righe = (
        sb.table(svc.INVII)
        .update({"stato": "errore", "motivo": "annullato dall'admin", "conclusa_at": svc.adesso_utc().isoformat()})
        .eq("id", invio_id).eq("stato", "richiesto").execute().data
    )
    if not righe:
        raise HTTPException(status_code=409, detail="Si annulla solo una richiesta che l'esecutore non ha ancora preso.")
    return righe[0]

File: apps/web/src/lib/invio-commercialista.ts
// Invio degli XML al commercialista: la logica pura della scheda admin
// (components/admin/invio-commercialista.tsx). Decide il worker e rifiuta il DB;
// qui solo cosa mostrare, con quali parole, e quali percorsi il proxy
// (app/api/admin/clienti/[id]/invio-commercialista/[[...percorso]]) puo' inoltrare.

export type StatoInvio =
  | "richiesto" | "in_corso" | "inviato" | "errore" | "bloccato" | "prova_ok" | "esito_incerto";
export type TipoInvio = "primo" | "ordinario" | "reinvio" | "prova";
export type Frequenza = "settimanale" | "quindicinale" | "mensile";
export type Tono = "positivo" | "negativo" | "incerto" | "neutro";

export interface Invio {
  id: string;
  tipo: TipoInvio;
  stato: StatoInvio;
  periodo_dal: string;
  periodo_al: string;
  richiesto_da: "notturno" | "admin";
  n_file: number | null;
  byte_totali: number | null;
  destinatario: string | null;
  motivo: string | null;
  creata_at: string;
  conclusa_at: string | null;
  email_tentata_at: string | null;
  link_scade_il: string | null;
  file_rimossi_at: string | null;
}

export interface Configurazione {
  id: string;
  piva: string;
  invoicetronic_company_id: number | null;
  invoicetronic_nome: string | null;
  email_destinatario: string | null;
  frequenza: Frequenza;
  data_partenza: string | null;
  attivo: boolean;
  consenso_ricevuto: boolean;
  consenso_data: string | null;
  consenso_email: string | null;
  sospesa_at: string | null;
  sospesa_motivo: string | null;
  aggiornata_at: string;
  sede_sdi_attiva: boolean;
  ultimo_giorno_inviato: string | null;
  invii: Invio[];
}

export interface StatoInvioCommercialista {
  oggi: string;
  limite_due_anni: string;
  piva_disponibili: string[];
  configurazioni: Configurazione[];
}

export const DUE_ANNI =
  "Invoicetronic conserva le fatture ricevute per 2 anni. Per periodi precedenti usa il Cassetto fiscale dell'Agenzia delle Entrate.";

export const ETICHETTA_STATO: Record<StatoInvio, string> = {
  richiesto: "In coda",
  in_corso: "In corso",
  inviato: "Inviato",
  errore: "Non riuscito",
  bloccato: "Bloccato",
  prova_ok: "Prova riuscita",
  esito_incerto: "Esito incerto",
};

export const TONO_STATO: Record<StatoInvio, Tono> = {
  richiesto: "neutro",
  in_corso: "neutro",
  inviato: "positivo",
  errore: "negativo",
  bloccato: "negativo",
  prova_ok: "positivo",
  esito_incerto: "incerto",
};

export const ETICHETTA_TIPO: Record<TipoInvio, string> = {
  primo: "Primo invio",
  ordinario: "Invio",
  reinvio: "Reinvio",
  prova: "Prova a vuoto",
};

export const ETICHETTA_FREQUENZA: Record<Frequenza, string> = {
  settimanale: "Settimanale (il lunedì)",
  quindicinale: "Quindicinale (l'1 e il 16)",
  mensile: "Mensile (l'1)",
};

// I motivi scritti dal worker (services/invio_commercialista_service.py) e i
// codici della guardia (services/invio_commercialista_guardia.py).
const MOTIVI: Record<string, string> = {
  invio_spento: "Interruttore generale spento (INVIO_COMMERCIALISTA_ATTIVO sul queue-worker): parte solo la prova a vuoto.",
  non_piu_autorizzato: "Dopo la richiesta la configurazione è stata spenta, sospesa o ha cambiato email.",
  configurazione_cambiata_durante_l_invio: "La configurazione è cambiata mentre si preparavano i file: nessuna email è partita.",
  configurazione_assente: "La configurazione non esiste più.",
  nessuna_sede_con_sdi_attivo: "Nessuna sede di questa P.IVA ha la ricezione SDI attiva.",
  brevo_non_configurato: "Manca BREVO_API_KEY sul queue-worker.",
  brevo_esito_incerto: "Brevo non ha confermato: l'email potrebbe essere partita. Controlla i log di Brevo e chiarisci.",
  interrotto: "Il worker si è fermato a metà: nessuna email è partita.",
  interrotto_dopo_l_email: "Il worker si è fermato dopo aver tentato l'email: controlla i log di Brevo e chiarisci.",
  riga_non_piu_in_corso: "L'invio è stato chiuso da un altro processo mentre lavorava.",
  saldo_invoicetronic_esaurito: "Saldo Invoicetronic esaurito durante i download.",
  saldo_insufficiente: "Saldo Invoicetronic troppo basso per scaricare tutto senza lasciare a secco le fatture in arrivo.",
  saldo_illeggibile: "Non si è riusciti a leggere il saldo Invoicetronic.",
  contatore_email_illeggibile: "Non si è riusciti a contare le email già inviate (anti-loop).",
  troppe_email_allo_stesso_destinatario_in_24_ore: "Già 3 email a questo indirizzo nelle ultime 24 ore (anti-loop).",
  zip_oltre_il_limite_del_bucket: "Un mese supera i 50 MB: lo ZIP non entra nel bucket.",
  invoicetronic: "Invoicetronic ha risposto con un errore.",
  errore_interno: "Errore interno del worker.",
  piva_non_solo_del_cliente: "La P.IVA risulta anche su un altro account.",
  azienda_assente_su_invoicetronic: "L'azienda non c'è più su Invoicetronic.",
  azienda_diversa_su_invoicetronic: "Invoicetronic risponde con un'azienda diversa da quella collegata.",
  documento_di_un_altra_azienda: "Nell'elenco c'è un documento di un'altra azienda.",
  documento_con_altro_destinatario: "Nell'elenco c'è un documento per un'altra P.IVA.",
  azienda_con_fatture_di_un_altro_cliente: "Fatture di questa azienda risultano arrivate a un altro cliente.",
  fatture_del_cliente_su_un_altra_azienda: "Fatture di questo cliente risultano arrivate su un'altra azienda Invoicetronic.",
  elenco_senza_fatture_gia_arrivate: "L'elenco di Invoicetronic non contiene fatture che OneFlux ha già ricevuto.",
  registro_diverso_dalla_configurazione: "L'invio non corrisponde alla sua configurazione.",
  stesso_nome_contenuto_diverso: "Due file con lo stesso nome e contenuto diverso.",
  stesso_identificativo_sdi_contenuto_diverso: "Due file con lo stesso identificativo SDI e contenuto diverso.",
};

/** Il motivo di un invio in una frase. Un codice sconosciuto resta com'è: meglio
 * il codice che una frase inventata. Il dettaglio fra parentesi o dopo i due
 * punti resta in coda. */
export function testoMotivo(motivo: string | null | undefined): string {
  if (!motivo) return "";
  const brevo = /^brevo_rifiutata_http_(\d{3})$/.exec(motivo);
  if (brevo) return `Brevo ha rifiutato l'email (HTTP ${brevo[1]}): ricontrolla l'indirizzo.`;
  const m = /^([a-z_]+)(?::\s*|\s+\(|$)(.*)$/.exec(motivo);
  if (!m || !(m[1] in MOTIVI)) return motivo;
  const dettaglio = m[2].replace(/\)$/, "").trim();
  return dettaglio ? `${MOTIVI[m[1]]} (${dettaglio})` : MOTIVI[m[1]];
}

/** Cosa manca perché il DB lasci accendere la configurazione (icc_attivabile_chk). */
export function requisitiMancanti(c: Configurazione): string[] {
  const mancano: string[] = [];
  if (c.invoicetronic_company_id == null) mancano.push("collegamento a Invoicetronic");
  if (!c.email_destinatario) mancano.push("email del commercialista");
  if (!c.consenso_ricevuto || !c.email_destinatario || c.consenso_email !== c.email_destinatario) {
    mancano.push("consenso per questa email");
  }
  if (!c.data_partenza) mancano.push("data di partenza");
  return mancano;
}

export function statoConfigurazione(c: Configurazione): { testo: string; tono: Tono } {
  if (c.sospesa_at) {
    return { testo: `Sospesa dalla guardia: ${testoMotivo(c.sospesa_motivo)}`, tono: "negativo" };
  }
  if (c.invii.some((i) => i.stato === "esito_incerto")) {
    return { testo: "Esito incerto da chiarire: nessun altro invio finché non lo chiudi", tono: "incerto" };
  }
  if (!c.attivo) {
    const mancano = requisitiMancanti(c);
    return { testo: mancano.length ? `Spenta — manca: ${mancano.join(", ")}` : "Spenta", tono: "neutro" };
  }
  if (!c.sede_sdi_attiva) return { testo: "In pausa: nessuna sede con SDI attivo", tono: "incerto" };
  if (!c.ultimo_giorno_inviato) return { testo: "Attiva — il primo invio si lancia a mano", tono: "incerto" };
  return { testo: `Attiva — inviato fino al ${formattaData(c.ultimo_giorno_inviato)}`, tono: "positivo" };
}

export function spostaGiorni(iso: string, giorni: number): string {
  const d = new Date(`${iso}T00:00:00Z`);
  d.setUTCDate(d.getUTCDate() + giorni);
  return d.toISOString().slice(0, 10);
}

/** Come `oggi - interval '2 years'` di Postgres: il 29/02 diventa il 28/02. */
export function limiteDueAnni(oggi: string): string {
  const anno = Number(oggi.slice(0, 4)) - 2;
  const meseGiorno = oggi.slice(5) === "02-29" ? "02-28" : oggi.slice(5);
  return `${String(anno).padStart(4, "0")}-${meseGiorno}`;
}

/** Lo stesso controllo del worker su un periodo scelto dall'admin (prova, reinvio). */
export function erroreDelPeriodo(dal: string, al: string, oggi: string): string | null {
  if (!dal || !al) return "Indica il periodo.";
  if (dal > al) return "La data di inizio viene dopo quella di fine.";
  if (al >= oggi) return "Il periodo deve finire al più tardi ieri.";
  if (dal < limiteDueAnni(oggi)) return DUE_ANNI;
  return null;
}

/** Il periodo che «Invia ora» chiederà: dalla data di partenza finché non c'è un
 * primo invio riuscito, poi dal giorno dopo l'ultimo inviato. Fino a ieri. */
export function periodoInviaOra(
  c: Pick<Configurazione, "data_partenza" | "ultimo_giorno_inviato">,
  oggi: string,
): { tipo: "primo" | "ordinario"; dal: string; al: string } | null {
  const limite = limiteDueAnni(oggi);
  const al = spostaGiorni(oggi, -1);
  let tipo: "primo" | "ordinario";
  let dal: string;
  if (c.ultimo_giorno_inviato) {
    tipo = "ordinario";
    dal = spostaGiorni(c.ultimo_giorno_inviato, 1);
  } else if (c.data_partenza) {
    tipo = "primo";
    dal = c.data_partenza;
  } else {
    return null;
  }
  if (dal < limite) dal = limite;
  return dal > al ? null : { tipo, dal, al };
}

export function formattaData(iso: string | null | undefined): string {
  if (!iso) return "—";
  const [a, m, g] = iso.slice(0, 10).split("-");
  return `${g}/${m}/${a}`;
}

export function formattaByte(n: number | null | undefined): string {
  if (n == null) return "—";
  if (n < 1024) return `${n} byte`;
  if (n < 1024 * 1024) return `${(n / 1024).toFixed(0)} KB`;
  return `${(n / (1024 * 1024)).toFixed(1).replace(".", ",")} MB`;
}

const ID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

/** Il percorso del worker per una richiesta al proxy, o null se non e' uno dei
 * percorsi di questa scheda. Il proxy non inoltra nient'altro: niente `..`,
 * niente segmenti arbitrari, e della query solo la P.IVA a 11 cifre. */
export function percorsoProxy(clienteId: string, segmenti: string[], query: URLSearchParams): string | null {
  if (!ID.test(clienteId)) return null;
  const base = `/api/admin/clienti/${clienteId}/invio-commercialista`;
  if (segmenti.length === 0) return base;
  if (segmenti.length === 1 && segmenti[0] === "azienda") {
    const piva = query.get("piva") ?? "";
    return /^\d{11}$/.test(piva) ? `${base}/azienda?piva=${piva}` : null;
  }
  const [config, ...resto] = segmenti;
  if (!ID.test(config)) return null;
  if (resto.length === 0) return `${base}/${config}`;
  if (resto[0] !== "invii") return null;
  if (resto.length === 1) return `${base}/${config}/invii`;
  if (resto.length === 3 && ID.test(resto[1]) && (resto[2] === "chiarisci" || resto[2] === "annulla")) {
    return `${base}/${config}/invii/${resto[1]}/${resto[2]}`;
  }
  return null;
}
